lib/semantics/scope.cc

```diff
@@ -265,27 +267,32 @@ const DeclTypeSpec &Scope::InstantiateIntrinsicType(
   // The expression was not originally constant, but now it must be so
   // in the context of a parameterized derived type instantiation.
   KindExpr copy{intrinsic->kind()};
+  evaluate::FoldingContext &foldingContext{semanticsContext.foldingContext()};
   copy = evaluate::Fold(foldingContext, std::move(copy));
-  auto value{evaluate::ToInt64(copy)};
-  CHECK(value.has_value() &&
-      "KIND parameter of intrinsic type did not resolve to a "
-      "constant INTEGER value in a parameterized derived type instance");
-  if (!evaluate::IsValidKindOfIntrinsicType(intrinsic->category(), *value)) {
-    foldingContext.messages.Say(
-        "KIND parameter value (%jd) of intrinsic type %s did not resolve to a supported value"_err_en_US,
-        static_cast<std::intmax_t>(*value),
-        parser::ToUpperCaseLetters(common::EnumToString(intrinsic->category()))
-            .data());
+  int kind{
+      semanticsContext.defaultKinds().GetDefaultKind(intrinsic->category())};
+  if (auto value{evaluate::ToInt64(copy)}) {
+    if (evaluate::IsValidKindOfIntrinsicType(intrinsic->category(), *value)) {
+      kind = *value;
+    } else {
+      foldingContext.messages().Say(
+          "KIND parameter value (%jd) of intrinsic type %s "
+          "did not resolve to a supported value"_err_en_US,
```

Comment:
When do we get this error as opposed to "... is not a supported type" from `semantics/expression.cc`?

Reply:
The error in `scope.cc` will be emitted only for a parameterized derived type component with an intrinsic type whose kind is an expression that depends on the kind type parameters of the PDT.   Resolution of those errors is necessarily deferred until the PDT is instantiated. The error in `expression.cc` will come out for all other bad cases of `KindSelector` as they are encountered in declarations, array constructors, and (soon) structure constructors and `ALLOCATE` statements.